cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java

```diff
@@ -0,0 +1,236 @@
+import com.google.gson.Gson;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.EventHandler;
+import org.apache.twill.api.EventHandlerContext;
+import org.apache.twill.api.RunId;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * A Twill {@link EventHandler} that responds to Twill application lifecycle events and aborts the application if
+ * it cannot provision container for some runnable.
+ */
+public class TwillAppLifecycleEventHandler extends AbortOnTimeoutEventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TwillAppLifecycleEventHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final String HADOOP_CONF_FILE_NAME = "hConf.xml";
+  private static final String CDAP_CONF_FILE_NAME = "cConf.xml";
+
+  private RunId twillRunId;
+  private ProgramRunId programRunId;
+  private ProgramStateWriter programStateWriter;
+  private ZKClientService zkClientService;
+  private AtomicBoolean launchedContainer;
+  private AtomicBoolean stoppedContainer;
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has no
+   * containers and a program state writer to respond to Twill application lifecycle events
+   *
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, ProgramOptions options) {
+    this(abortTime, false, options);
+  }
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has not enough
+   * containers.
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param abortIfNotFull If {@code true}, it will abort the application if any runnable doesn't meet the expected
+   *                       number of instances.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, boolean abortIfNotFull, ProgramOptions options) {
+    super(abortTime, abortIfNotFull);
+    this.programRunId = options.getProgramId().run(ProgramRunners.getRunId(options));
+  }
+
+  @Override
+  protected Map<String, String> getConfigs() {
+    Map<String, String> configs = super.getConfigs();
+    configs.put("programRunId", GSON.toJson(programRunId));
```

Comment:
The map returned by parent is immutable. This this throw exception. It's better to do
`Map<String, String> configs = new HashMapM<>(super.getConfigs());`